app/controllers/find/candidates/saved_courses_controller.rb

```diff
@@ -17,12 +17,36 @@ def create
         end
       end
 
+      def undo
+        course = Course.find(params[:course_id])
+
+        if SaveCourseService.call(candidate: @candidate, course:)
+          redirect_to find_candidate_saved_courses_path
+        else
+          redirect_to find_candidate_saved_courses_path(course, error: t(".save_failed"))
+        end
+      end
+
       def destroy
         saved_course = @candidate.saved_courses.find(params[:id])
         course = Course.find(saved_course.course_id)
 
         if saved_course.destroy
-          redirect_to_course(course)
+          undo_link = view_context.form_with url: undo_find_candidate_saved_courses_path, method: :post do |form|
+            form.hidden_field(:course_id, value: course.id) + form.submit(t(".undo"), class: "app-button-link govuk-!-padding-left-0")
+          end
```

Comment:
In my personal opinion, I think is best to not manipulate HTML in the controller.

Suggestion 1: Component:

```ruby
undo_link = view_context.render(
  UndoLinkComponent.new(
    label: t(".undo"),
    path: undo_find_candidate_saved_courses_path,
    course:
  )
)
```

Suggestion 2: Partial

```ruby
undo_link = view_context.render(
  partial: "saved_courses/undo_link",
  locals: { label: t(".undo"), undo_path: undo_find_candidate_saved_courses_path, course: }
)
```

Suggestion 3: Helper method.